lib/spack/spack/util/compression.py

```diff
@@ -201,6 +199,40 @@ def _lzma_decomp(archive_file):
             return _xz(archive_file)
 
 
+def _win_compressed_tarball_handler(archive_file):
+    """Decompress and extract compressed tarballs on Windows.
+    This method uses 7zip in conjunction with the tar utility
+    to perform decompression and extraction in a two step process
+    first using 7zip to decompress, and tar to extract.
+
+    The motivation for this method is the inability of 7zip
+    to directly decompress and extract compressed archives
+    in a single shot without undocumented workarounds, and
+    the Windows tar utility's lack of access to the xz tool (unsupported on Windows)
+    """
+    # perform intermediate extraction step
+    # record name of new archive so we can extract
+    # and later clean up
+    decomped_tarball = _7zip(archive_file)
+    # 7zip is able to one shot extract compressed archives
+    # that have been named .txz. If that is the case, there will
+    # be no intermediate archvie to extract.
+    if os.path.exists(decomped_tarball):
+        # run tar on newly decomped archive
+        outfile = _untar(decomped_tarball)
+        # clean intermediate archive to mimic end result
+        # produced by one shot decomp/extraction
+        os.remove(decomped_tarball)
+        return outfile
+    else:
+        # File was a .txz, 7zip can directly extract/decompress these
+        # in one shot (this behavior is undocumented)
+        # so be sure we return the correct extensions in the path
+        _, ext = os.path.splitext(decomped_tarball)
+        decomped_tarball = os.path.basename(decomped_tarball.strip(ext))
```

Comment:
To remove a string off the end of another string: don't use `.strip()`: it's semantics are weird and will lead to unexpected scenarios. Use something like `re.sub(` (as is done in `strip_extension` - in fact if it's possible to consolidate it that would be good). I apparently missed that when reviewing the additions of the decompression methods like 7zip (that can be left as a TODO for later but I think it would be good to fix this one).

Another refactor (for later - but if you can add a TODO): I think `_7zip()` returning a file path that doesn't exist is a confusing condition to depend on in terms of deciding that what we should actually be doing is untar-ing the result.

Reply:
Agreed on the first point, will refactor. As to the second point, _7zip returning a path that doesn't exist is not ideal, and that should be refactored (although it will include some string parsing to determine if a file has an abbreviated extension) but that will still require a check in this method (again involving some semi-fragile string parsing) as well.

Reply:
I agree that regardless of how `_7zip` itself is factored that this will need a check. I just think that this arrangement should be less confusing than "the file path doesn't exist so it must be a .tar" (this part can be handled in a later PR though).